sdk/eventgrid/azure-messaging-eventgrid/src/main/java/com/azure/messaging/eventgrid/CloudEvent.java

```diff
@@ -0,0 +1,325 @@
+        }
+
+        this.cloudEvent = new com.azure.messaging.eventgrid.implementation.models.CloudEvent()
+            .setId(UUID.randomUUID().toString())
+            .setSource(source)
+            .setType(type)
+            .setSpecversion(SPEC_VERSION);
+    }
+
+    /**
+     * Parse the Cloud Event from a JSON string. This can be used to interpret the event at the event destination
+     * from raw JSON into rich event(s).
+     * @param json the JSON payload containing one or more events.
+     *
+     * @return all of the events in the payload parsed as CloudEvents.
+     */
+    public static List<CloudEvent> parse(String json) {
+        return Flux.fromArray(deserializer
+            .deserialize(new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8)),
+                TypeReference.createInstance(com.azure.messaging.eventgrid.implementation.models.CloudEvent[].class))
+        )
+            .map(event1 -> {
+                if (event1.getData() != null) {
+                    ByteArrayOutputStream stream = new ByteArrayOutputStream();
+                    deserializer.serialize(stream, event1.getData());
+                    return new CloudEvent(event1).setData(stream.toByteArray()); // use BinaryData instead?
+                } else { // both null, don't set data and keep null
+                    return new CloudEvent(event1);
+                }
+            })
+            .collectList()
+            .block();
+    }
+
+    /**
+     * Get the id of the cloud event.
+     * @return the id.
+     */
+    public String getId() {
+        return this.cloudEvent.getId();
+    }
+
+    /**
+     * Set a custom id. Note that a random id is already set by default.
+     * @param id the id to set.
+     *
+     * @return the cloud event itself.
+     */
+    public CloudEvent setId(String id) {
+        if (CoreUtils.isNullOrEmpty(id)) {
+            throw new IllegalArgumentException("id cannot be null or empty");
+        }
+        this.cloudEvent.setId(id);
+        return this;
+    }
+
+    /**
+     * Get the URI source of the event.
+     * @return the source.
+     */
+    public String getSource() {
+        return this.cloudEvent.getSource();
+    }
+
+    /**
+     * Get the data associated with this event. For use in a parsed event only.
+     * @return If the event was parsed from a Json, this method will return the rich
+     * system event data if it is a system event, and a {@code byte[]} otherwise, such as in the case of custom event
+     * data, including data set through {@link CloudEvent#setData(byte[], String)}.
+     * @throws IllegalStateException If the event was not created through {@link EventGridEvent#parse(String)}.
+     */
+    public Object getData() {
+        if (!parsed) {
+            // data was set instead of parsed, throw error
+            throw new IllegalStateException("This method should only be called on events created through the parse method");
```

Comment:
I think Srikanta already called this out, but instead of throwing directly, we should get a `ClientLogger` instance and call `throw logger.logExceptionAsWarning/Error(...)` instead.